model_signing/serialization/itemized.py

```diff
@@ -0,0 +1,228 @@
+from model_signing.serialization import dfs
+from model_signing.serialization import serialization
+
+
+class FilesSerializer(serialization.Serializer):
+    """Model serializers that produces an itemized manifest, at file level.
+
+    Traverses the model directory and creates digests for every file found,
+    possibly in parallel.
+
+    Since the manifest lists each item individually, this will also enable
+    support for incremental updates (to be added later).
+    """
+
+    def __init__(
+        self,
+        hasher_factory: Callable[[pathlib.Path], file.FileHasher],
+        max_workers: int | None = None,
+    ):
+        """Initializes an instance to serialize a model with this serializer.
+
+        Args:
+            hasher_factory: A callable to build the hash engine used to hash
+              individual files.
+            max_workers: Maximum number of workers to use in parallel. Default
+              is to defer to the `concurent.futures` library.
+        """
+        self._hasher_factory = hasher_factory
+        self._max_workers = max_workers
+
+    @override
+    def serialize(self, model_path: pathlib.Path) -> manifest.FileLevelManifest:
+        # TODO: github.com/sigstore/model-transparency/issues/196 - Add checks
+        # to exclude symlinks if desired.
+        dfs.check_file_or_directory(model_path)
+
+        paths = []
+        if model_path.is_file():
+            paths.append(model_path)
+        else:
+            # TODO: github.com/sigstore/model-transparency/issues/200 - When
+            # Python3.12 is the minimum supported version, this can be replaced
+            # with `pathlib.Path.walk` for a clearer interface, and some speed
+            # improvement.
+            for path in model_path.glob("**/*"):
+                dfs.check_file_or_directory(path)
+                if path.is_file():
+                    paths.append(path)
+
+        manifest_items = []
+        with concurrent.futures.ThreadPoolExecutor(
+            max_workers=self._max_workers
+        ) as tpe:
+            futures = [
+                tpe.submit(self._compute_hash, model_path, path)
+                for path in paths
+            ]
+            for future in concurrent.futures.as_completed(futures):
+                manifest_items.append(future.result())
+
+        return self._build_manifest(manifest_items)
+
+    def _compute_hash(
+        self, model_path: pathlib.Path, path: pathlib.Path
+    ) -> manifest.FileManifestItem:
+        """Produces the manifest item of the file given by `path`.
+
+        Args:
+            model_path: The path to the model.
+            path: Path to the file in the model, that is currently transformed
+              to a manifest item.
+
+        Returns:
+            The itemized manifest.
+        """
+        # Note: we need to force a PosixPath to canonicalize the manifest
+        relative_path = pathlib.PurePosixPath(path.relative_to(model_path))
```

Comment:
It might make sense to move this into the manifest class itself. So, a user cannot mishandle it.  Perhaps adding a custom init method to the dataclasses that take `path` and `model_path` as arguments and then makes the call to `PurePosixPath`.

Reply:
Oh, this is good idea, thank you. I'll try it once I get the unit tests working (need to convert the paths to Posix there too, after I alter the model)

Reply:
Actually, to do so we would have to unfreeze the dataclass. Let me think a little bit more if that's desired.